compiler/src/iree/compiler/Codegen/Common/RemoveTrivialLoops.cpp

```diff
@@ -104,32 +104,39 @@ static SmallVector<int64_t> getNumWorkgroup(
   auto allLoops = getTiledAndDistributedLoopInfo(funcOp);
   auto wgLoops =
       llvm::to_vector<3>(llvm::make_filter_range(allLoops, isWorkgroupLoop));
-  SmallVector<int64_t> workloadSize(wgLoops.size());
-  for (LoopTilingAndDistributionInfo &tileInfo : wgLoops) {
+
+  auto translationInfo = getTranslationInfo(exportOp);
+  if (!translationInfo) return SmallVector<int64_t>();
+  SmallVector<int64_t> workloadPerWorkgroup =
```

Comment:
The workloadPerWorkgroup was already used in this function. My change is to consider the case where wgLoops includes all the tiled loops. (Now we can have two nested loops.) In such case, my code checks whether the workload size already set by the first loop nest is the same as the second loop nest.

I am happy to enhance the code if there is any subtlety. What would you recommend?

Reply:
AFAIK, this isnt used, but is set in this pass cause it is needed later. This is used by the `Remove*SingleIteration*LoopPass` that uses this to remove the distributed loops. I really want to deprecate that pass. The way I was thinking about this was that you dont generate the loop to begin with. There is actually no need to generate the loop during distribution only to remove it. We could just not generate the loop to start with. This is what I was going to try.

Another reason why I am thinking twice about this change. If the loop doesnt exist to begin with, this approach wont work....

Reply:
(Didnt see that this change is to `RemoveTrivialLoops`, but the question above still holds)

Reply:
Using `scf.foreach_thread` would be a great way to remove the nested loops. My transformation should be easy to modify according to the plan. Let me know when it happens, I can help change the current transformation.

Reply:
Let me handle the static case removal by canonicalization in the next PR with dynamic shape support, which is coming in a couple of weeks.

For the second point, finding an IR for dynamic size seems inevitable even when we use `scf.foreach_thread`. We can introduce an op for the dynamic size calculation or we can attach a simple attribute to mark. Do you have a suggestion?

Reply:
Also, having an option is desirable because it can be a trade off between the code size and speed.

Reply:
> > Wanted some clarification on what your goal in the next few days are for this PR. Is this WIP or do you want to land this?
> 
> Wanna land it and continue working on it for performance study and dynamic shapes support. It is okay to make it disabled by default, but I don't see a huge regression from the benchmark suites.

Overall I see where this is going, and it make sense for the current state of things, but I am not sure it composes as well based on next steps here (and introducing more ops here is not a great idea. One of the things with MLIR is that adding ops is easy to get past immediate needs, but often they dont compose well with transformations)
Some next steps that I see becoming more involved

1) Currently the tile and distribute generates `scf.for` operations with distribution. Then the `workload_per_workgroup` is used to remove this loops by having a complex analysis to figure out that these are actually a single trip loops on each thread block. There is an option in distribution to not generate the loop to begin with (when it is known that the number of threads blocks is guaranteed to be equal to the number of tiles, which is the case here). Using this option  gives two things
   a) We dont need to complex analysis for removing the loop. That pass can be deprecated.
   b) We can remove `workload_per_workgroup` completely (its only kept around for the analysis in (a) and has some nasty side-effects.
This was one of the main reasons I refactored this code. If we go down this path, the PR here becomes defunct. How do you make this work if there are no loops to begin with?
2) An alternative to the above is to use `scf.foreach_thread` which has the advantage of using things completely from upstream. So there will be very little "IREE specific" transformations in the backends. Even this route this PR becomes defunct.

I am deciding between (1) and (2), both are unfinished, but (2) is more long term stable. Either way, the approach in this PR cannot be supported with those.

I see there is a lot of analysis here w.r.t `affine.min`. That is an artifact of the code-generation being conservative. What you are really looking for is to peel out the full tiles from the partial tiles. So there are two approaches
1) If you know that you have full tiles and partial tiles, then generate the full tile code and partial tile code directly.
2) Just generate the full tile and partial tile code by default, and then rely on canonicalizations to constant propagate (like removing affine.min ops, or removing the partial tile code if we have only full tiles).

If I were aware from the start the direction being taken here, I would have advised against going down this path for the reasons above. I would very much have liked some help doing either (1) or (2) instead....

Reply:
I see your concern with affine.min. I am open to change the code after we make a transition to scf.foreach_thread. Even further, I can wait for it to happen without merging this PR if it is going to happen soon and that's what you want. I don't have a very specific flavor in terms of how it is implemented. (I took the simplest way in terms of implementation.)

> I see there is a lot of analysis here w.r.t affine.min

To be clear, there isn't a lot of analysis here. Actually there is no analysis regarding the affine.min; I just replaced the bounded size ops with the tile size and the bounded size ops are collected when creating the loop nest.

Again, on top of the current distribution loop, collecting the bounded size ops and replacing them with the tile size seems to be the simplest thing to do.

> If you know that you have full tiles and partial tiles, then generate the full tile code and partial tile code directly.

I will take a look at the approach to see if it can be done without depending the generated loops (whether it is scf.for or scf.foreach_thread). Thanks for the comment.